mgplvm/rdist/rGP.py

```diff
@@ -0,0 +1,421 @@
+        #print('gmu, gamma, x shapes', gmu.shape, gamma.shape, x.shape)
+
+        if analytic_kl:
+            lq = self.kl(prior, gmu.transpose(-1, -2), gamma,
+                         Lq)  #(n_samples x d)
+        else:
+            lq = q.log_prob(x.transpose(-1, -2))  #(n_mc x n_samples x d)
+            #print('lq shape:', lq.shape)
+            #sum across d
+            lq = lq.sum(-1)[..., None]  #(n_mc x n_samples x 1)
+
+        #print('x, mu shapes', x.shape, gmu.shape)
+        x = x + gmu[None, ...]  #add mean
+        return x, lq
+
+    def kl(self, prior, gmu, gamma_q, Lq):
+        """
+        compute KL divergence analytically
+        gamma_q is (n_samples x d x m x m)
+        gmu is (n_samples x d x m)
+        """
+        gamma_p = prior.prms  #(n_samples x d x m x m)
+        Lp = torch.cholesky(gamma_p)
+        muq = gmu
+        m = Lp.shape[-1]  #dimensionality of the Gaussian
+        sqTerm = torch.square(muq[..., None, :] @ torch.inverse(Lp)).sum(
+            (-1, -2))  #(n_samples x d)
+        logDetq = 2 * torch.log(torch.diagonal(Lq, dim1=-1, dim2=-2)).sum(
+            -1)  # log |S1| (n_samples x d)
+        logDetp = 2 * torch.log(torch.diagonal(Lp, dim1=-1, dim2=-2)).sum(
+            -1)  # log |S2|
+        cholsolve = torch.cholesky_solve(gamma_q, Lp)
+        trterm = torch.diagonal(cholsolve, dim1=-1,
+                                dim2=-2).sum(-1)  #(n_samples x d)
+        kl = 0.5 * (trterm + sqTerm - m + logDetp - logDetq)  #(n_samples x d)
+
+        return kl
+
+    def gmu_parameters(self):
+        return self.f.gmu_parameters()
+
+    def concentration_parameters(self):
+        return self.f.concentration_parameters()
+
+
+class _F(Module):
+
+    def __init__(self,
+                 manif: Manifold,
+                 m: int,
+                 n_samples: int,
+                 ts: Tensor,
+                 mu=None,
+                 initialization: Optional[str] = 'random',
+                 Y=None):
+
+        super(_F, self).__init__()
+        self.manif = manif
+        self.d = manif.d
+
+        #initialize GP mean
+        if mu is None:
+            gmu = self.manif.initialize(initialization, n_samples, m, manif.d,
+                                        Y)
+        else:
+            assert mu.shape == (n_samples, m, manif.d2)
+            gmu = torch.tensor(mu)
+        self.gmu = nn.Parameter(data=gmu, requires_grad=True)
+
+        #initialize covariance parameters
+        jitter = 1e-3
+        #_scale = torch.ones(n_samples, self.d, m)*0.5
+        _scale = torch.ones(1, self.d, 1) * 0.5
+        ell = (torch.max(ts) - torch.min(ts)) / 10
+        _ell = torch.ones(1, self.d, 1, 1) * ell
+
+        self._scale = nn.Parameter(
+            data=inv_softplus(_scale), requires_grad=True
+        )  #can add separate scale for each latent in the future
+        self._ell = nn.Parameter(data=inv_softplus(_ell), requires_grad=True)
+        self.noise = torch.ones(
```

Comment:
this doesn't seem to be used? @KrisJensen

Reply:
not sure what you mean?

Reply:
`self.noise` is defined twice here, not sure which one you mean

Reply:
ahh I see -- I use the second (scalar) one, but we will probably end up culling this whole class and go with the more efficient parameterization at some point in the future anyways.

Reply:
okay i'll just comment out the first line for now